x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java

```diff
@@ -828,6 +842,72 @@ public void testMetadataIsNotSentToClient() {
+        assertThat(authenticateResponse.authentication().getLookedUpBy().getName(), equalTo("index"));
+        assertThat(authenticateResponse.authentication().getAuthenticatedBy().getName(), equalTo("file"));
+        assertThat(authenticateResponse.authentication().getAuthenticationType(), is(Authentication.AuthenticationType.TOKEN));
+        // Invalidate tokens by realm and username respect the run-as user's username and realm
+        var invalidateTokenRequest = new org.elasticsearch.xpack.core.security.action.token.InvalidateTokenRequest(
+            null,
+            null,
+            "index",
+            nativeTokenUsername
+        );
+        final PlainActionFuture<org.elasticsearch.xpack.core.security.action.token.InvalidateTokenResponse> future2 =
+            new PlainActionFuture<>();
+        client().execute(InvalidateTokenAction.INSTANCE, invalidateTokenRequest, future2);
+        assertThat(future2.actionGet().getResult().getInvalidatedTokens().size(), equalTo(1));
+
+        // Create a token with password grant and run-as user (native realm)
+        var createTokenRequest2 = new org.elasticsearch.xpack.core.security.action.token.CreateTokenRequest();
+        createTokenRequest2.setGrantType("password");
+        createTokenRequest2.setUsername(ES_TEST_ROOT_USER);
+        createTokenRequest2.setPassword(TEST_PASSWORD_SECURE_STRING.clone());
+        final PlainActionFuture<org.elasticsearch.xpack.core.security.action.token.CreateTokenResponse> future3 = new PlainActionFuture<>();
+        runAsClient.execute(CreateTokenAction.INSTANCE, createTokenRequest2, future3);
+        final String refreshToken = future3.actionGet().getRefreshToken();
+
+        var createTokenRequest3 = new org.elasticsearch.xpack.core.security.action.token.CreateTokenRequest();
+        createTokenRequest3.setGrantType("refresh_token");
+        createTokenRequest3.setRefreshToken(refreshToken);
+        final PlainActionFuture<org.elasticsearch.xpack.core.security.action.token.CreateTokenResponse> future4 = new PlainActionFuture<>();
+
+        // Refresh token is bound to the original user that creates it. In this case, it is the run-as user
+        // refresh without run-as should fail
+        client().filterWithHeader(
+            Map.of("Authorization", UsernamePasswordToken.basicAuthHeaderValue(ES_TEST_ROOT_USER, TEST_PASSWORD_SECURE_STRING.clone()))
+        ).execute(RefreshTokenAction.INSTANCE, createTokenRequest3, future4);
+        expectThrows(ElasticsearchSecurityException.class, future4::actionGet);
+
+        // refresh with run-as should work
+        final PlainActionFuture<org.elasticsearch.xpack.core.security.action.token.CreateTokenResponse> future5 = new PlainActionFuture<>();
+        runAsClient.execute(RefreshTokenAction.INSTANCE, createTokenRequest3, future5);
+        assertThat(future5.actionGet().getTokenString(), notNullValue());
```

Comment:
I am keeping this test since existing ones do not cover using refresh_token in this scenario ("user1 run-as user2 and creates token for user3").